When adding a db, immediately updating `database-enable-actions` prevents the db from finishing initial sync

I discovered this while setting up automated testing for actions DBs, If we update the DB immediately after adding it, before the initial sync had finished, the DB would always report that the initial sync was "incomplete"

This was an easy workaround in our e2e test setup: wait for the initial sync to complete before enabling actions, but I fear that real customers with large DBs that take minutes to sync initially could get stuck.

I finally able to reproduce this with the dev server.

The error is from and it's failing to get the connection:
https://github.com/metabase/metabase/blob/e896c3ddfc21d358e29d2ca31530253eb718bf70/src/metabase/driver/sql_jdbc/sync/describe_database.clj#L148

Error: ` java.sql.SQLException: An SQLException was provoked by the following failure: java.lang.InterruptedException`

<details>
<summary>stacktrace </summary>

```

 java.sql.SQLException: An SQLException was provoked by the following failure: java.lang.InterruptedException
        at com.mchange.v2.sql.SqlUtils.toSQLException(SqlUtils.java:118)
        at com.mchange.v2.sql.SqlUtils.toSQLException(SqlUtils.java:77)
        at com.mchange.v2.sql.SqlUtils.toSQLException(SqlUtils.java:74)
        at com.mchange.v2.c3p0.impl.C3P0PooledConnectionPool.checkoutPooledConnection(C3P0PooledConnectionPool.java:694)
        at com.mchange.v2.c3p0.impl.AbstractPoolBackedDataSource.getConnection(AbstractPoolBackedDataSource.java:140)
        at clojure.java.jdbc$get_connection.invokeStatic(jdbc.clj:372)
        at clojure.java.jdbc$get_connection.invoke(jdbc.clj:274)
        at clojure.java.jdbc$get_connection.invokeStatic(jdbc.clj:343)
        at clojure.java.jdbc$get_connection.invoke(jdbc.clj:274)
        at metabase.driver.sql_jdbc.sync.describe_database$describe_database.invokeStatic(describe_database.clj:153)
        at metabase.driver.sql_jdbc.sync.describe_database$describe_database.invoke(describe_database.clj:148)
        at metabase.driver.sql_jdbc$eval109547$fn__109548.invoke(sql_jdbc.clj:71)
        at clojure.lang.MultiFn.invoke(MultiFn.java:234)
        at metabase.sync.fetch_metadata$eval113283$db_metadata__113288$fn__113289.invoke(fetch_metadata.clj:15)
        at metabase.sync.fetch_metadata$eval113283$db_metadata__113288.invoke(fetch_metadata.clj:12)
        at metabase.sync.sync_metadata$eval115103$sync_db_metadata_BANG___115108$fn__115109.invoke(sync_metadata.clj:58)
        at metabase.sync.sync_metadata$eval115103$sync_db_metadata_BANG___115108.invoke(sync_metadata.clj:55)
        at metabase.sync$eval115162$sync_database_BANG___115171$fn__115177$fn__115179$fn__115181.invoke(sync.clj:47)
        at clojure.core$mapv$fn__8535.invoke(core.clj:6979)
        at clojure.lang.ArrayChunk.reduce(ArrayChunk.java:58)
        at clojure.core.protocols$fn__8244.invokeStatic(protocols.clj:136)
        at clojure.core.protocols$fn__8244.invoke(protocols.clj:124)
        at clojure.core.protocols$fn__8204$G__8199__8213.invoke(protocols.clj:19)
        at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:31)
        at clojure.core.protocols$fn__8236.invokeStatic(protocols.clj:75)
        at clojure.core.protocols$fn__8236.invoke(protocols.clj:75)
        at clojure.core.protocols$fn__8178$G__8173__8191.invoke(protocols.clj:13)
        at clojure.core$reduce.invokeStatic(core.clj:6886)
        at clojure.core$mapv.invokeStatic(core.clj:6970)
        at clojure.core$mapv.invoke(core.clj:6970)
        at metabase.sync$eval115162$sync_database_BANG___115171$fn__115177$fn__115179.invoke(sync.clj:47)
        at metabase.sync.util$do_with_error_handling.invokeStatic(util.clj:159)
        at metabase.sync.util$do_with_error_handling.invoke(util.clj:152)
        at clojure.core$partial$fn__5910.invoke(core.clj:2647)
        at metabase.driver$eval72920$fn__72921.invoke(driver.clj:662)
        at clojure.lang.MultiFn.invoke(MultiFn.java:239)
        at metabase.sync.util$sync_in_context$fn__82603.invoke(util.clj:137)
        at metabase.sync.util$with_db_logging_disabled$fn__82600.invoke(util.clj:128)
        at metabase.sync.util$with_start_and_finish_logging_STAR_.invokeStatic(util.clj:101)
        at metabase.sync.util$with_start_and_finish_logging_STAR_.invoke(util.clj:95)
        at metabase.sync.util$with_start_and_finish_logging$fn__82589.invoke(util.clj:113)
        at metabase.sync.util$with_sync_events$fn__82584.invoke(util.clj:87)
        at metabase.sync.util$with_duplicate_ops_prevented$fn__82575.invoke(util.clj:66)
        at metabase.sync.util$do_sync_operation.invokeStatic(util.clj:180)
        at metabase.sync.util$do_sync_operation.invoke(util.clj:177)
        at metabase.sync$eval115162$sync_database_BANG___115171$fn__115177.invoke(sync.clj:46)
        at metabase.sync$eval115162$sync_database_BANG___115171.invoke(sync.clj:32)
        at metabase.sync$eval115162$sync_database_BANG___115171$fn__115172.invoke(sync.clj:42)
        at metabase.sync$eval115162$sync_database_BANG___115171.invoke(sync.clj:32)
        at metabase.events.sync_database$process_sync_database_event$fn__144799.invoke(sync_database.clj:39)
        at clojure.core$binding_conveyor_fn$fn__5823.invoke(core.clj:2047)
        at clojure.lang.AFn.call(AFn.java:18)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.InterruptedException
        at java.base/java.lang.Object.wait(Native Method)
        at com.mchange.v2.resourcepool.BasicResourcePool.awaitAvailable(BasicResourcePool.java:1503)
        at com.mchange.v2.resourcepool.BasicResourcePool.prelimCheckoutResource(BasicResourcePool.java:644)
        at com.mchange.v2.resourcepool.BasicResourcePool.checkoutResource(BasicResourcePool.java:554)
        at com.mchange.v2.c3p0.impl.C3P0PooledConnectionPool.checkoutAndMarkConnectionInUse(C3P0PooledConnectionPool.java:758)
        at com.mchange.v2.c3p0.impl.C3P0PooledConnectionPool.checkoutPooledConnection(C3P0PooledConnectionPool.java:685)
        ... 52 more

```
</details>


I can *sometimes* reproduce with this script. I think it is a race condition problem

```clojure
(let [do-n-times (fn [n sleep thunk]
                   (doseq [_ (range n)]
                     (thunk)
                     (Thread/sleep sleep)))
      db (metabase.test/user-http-request :crowberto :post 200 "/database"
                                          {:is_on_demand false
                                           :is_full_sync true
                                           :is_sample false
                                           :cache_ttl nil
                                           :refingerprint false
                                           :auto_run_queries true
                                           :schedules {}
                                           :details {:host "localhost"
                                                     :port 5404
                                                     :dbname "sample"
                                                     :user "metabase"
                                                     :password "metasample123"
                                                     :schema-filters-type :all
                                                     :ssl false
                                                     :tunnel-enabled false
                                                     :advanced-options false}
                                           :name   "test sync"
                                           :engine "postgres"})]
   (metabase.test/user-http-request :crowberto :put 200 (format "/database/%d" (:id db))
                                    {:settings {"database-enable-actions" true}})
   (do-n-times 3 1000 (fn [] (println (:initial_sync_status (metabase.test/user-http-request :crowberto :get 200 (format "/database/%d" (:id db))))))))
```

--- a/src/metabase/events/driver_notifications.clj
+++ b/src/metabase/events/driver_notifications.clj
@@ -5,6 +5,7 @@
   this writing, the SQL JDBC driver 'superclass' is the only thing that implements this method, and does so to close
   connection pools when database details change or when they are deleted."
   (:require
+   [medley.core :as m]
    [metabase.driver :as driver]
    [metabase.events :as events]
    [metabase.util.log :as log]
@@ -15,10 +16,15 @@
 (derive :event/database-delete ::event)
 
 (methodical/defmethod events/publish-event! ::event
-  [topic {database :object :as _event}]
+  [topic {database :object, previous-database :previous-object :as _event}]
   ;; try/catch here to prevent individual topic processing exceptions from bubbling up.  better to handle them here.
   (try
-   ;; notify the appropriate driver about the updated database
-   (driver/notify-database-updated (:engine database) database)
-   (catch Throwable e
-     (log/warnf e "Failed to process driver notifications event. %s" topic))))
+    ;; notify the appropriate driver about the updated database to release any resources it may be holding onto.
+    ;; don't notify the driver if the changes should not impact the state of any resource.
+    (let [paths-that-dont-impact-resource-state [[:updated_at]
+                                                 [:settings :database-enable-actions]]]
+      (when (not= (reduce m/dissoc-in database paths-that-dont-impact-resource-state)
+                  (reduce m/dissoc-in previous-database paths-that-dont-impact-resource-state))
+        (driver/notify-database-updated (:engine database) database)))
+    (catch Throwable e
+      (log/warnf e "Failed to process driver notifications event. %s" topic))))

--- a/src/metabase/events/driver_notifications.clj
+++ b/src/metabase/events/driver_notifications.clj
@@ -19,12 +19,15 @@
   [topic {database :object, previous-database :previous-object :as _event}]
   ;; try/catch here to prevent individual topic processing exceptions from bubbling up.  better to handle them here.
   (try
-    ;; notify the appropriate driver about the updated database to release any resources it may be holding onto.
-    ;; don't notify the driver if the changes should not impact the state of any resource.
-    (let [paths-that-dont-impact-resource-state [[:updated_at]
-                                                 [:settings :database-enable-actions]]]
-      (when (not= (reduce m/dissoc-in database paths-that-dont-impact-resource-state)
-                  (reduce m/dissoc-in previous-database paths-that-dont-impact-resource-state))
+    ;; notify the appropriate driver about the updated database to release any related resources, such as connections.
+    ;; avoid notifying if the changes shouldn't impact the observable behaviour of any resource, otherwise drivers might
+    ;; close connections or other resources unnecessarily (metabase#27877).
+    (let [;; remove data that should not impact the observable state of any resource before comparing
+          remove-irrelevant-data (fn [db]
+                                   (reduce m/dissoc-in db [[:updated_at]
+                                                           [:settings :database-enable-actions]]))]
+      (when (not= (remove-irrelevant-data database)
+                  (remove-irrelevant-data previous-database))
         (driver/notify-database-updated (:engine database) database)))
     (catch Throwable e
       (log/warnf e "Failed to process driver notifications event. %s" topic))))

--- a/test/metabase/api/database_test.clj
+++ b/test/metabase/api/database_test.clj
@@ -9,6 +9,7 @@
    [metabase.api.table :as api.table]
    [metabase.driver :as driver]
    [metabase.driver.h2 :as h2]
+   [metabase.driver.sql-jdbc.execute :as sql-jdbc.execute]
    [metabase.driver.util :as driver.u]
    [metabase.lib.schema.id :as lib.schema.id]
    [metabase.models
@@ -19,6 +20,7 @@
    [metabase.models.permissions-group :as perms-group]
    [metabase.models.setting :as setting :refer [defsetting]]
    [metabase.public-settings.premium-features :as premium-features]
+   [metabase.sync :as sync]
    [metabase.sync.analyze :as analyze]
    [metabase.sync.field-values :as field-values]
    [metabase.sync.sync-metadata :as sync-metadata]
@@ -33,7 +35,11 @@
    [metabase.util.malli.schema :as ms]
    [ring.util.codec :as codec]
    [toucan2.core :as t2]
-   [toucan2.tools.with-temp :as t2.with-temp]))
+   [toucan2.tools.with-temp :as t2.with-temp])
+  (:import
+   (java.sql Connection)))
+
+(set! *warn-on-reflection* true)
 
 (use-fixtures :once (fixtures/initialize :db :plugins :test-drivers))
 
@@ -510,6 +516,42 @@
                                                (format "database/%s" db-id)
                                                {:settings {:database-enable-actions true}})
                          [:settings :database-enable-actions]))))))
+
+(deftest update-database-enable-actions-open-connection-test
+  (testing "Updating a database's `database-enable-actions` setting shouldn't close existing connections (metabase#27877)"
+    (mt/test-drivers (filter #(isa? driver/hierarchy % :sql-jdbc) (mt/normal-drivers-with-feature :actions))
+      (let [;; 1. create a database and sync
+            database-name      (name (gensym))
+            empty-dbdef        {:database-name database-name}
+            _                  (tx/create-db! :postgres empty-dbdef)
+            connection-details (tx/dbdef->connection-details driver/*driver* :db empty-dbdef)
+            db                 (first (t2/insert-returning-instances! :model/Database {:name    database-name
+                                                                                       :engine  (u/qualified-name driver/*driver*)
+                                                                                       :details connection-details}))
+            _                  (sync/sync-database! db)]
+        (let [;; 2. start a long running process on another thread that uses a connection
+              connections-stay-open? (future
+                                       (sql-jdbc.execute/do-with-connection-with-options
+                                        driver/*driver*
+                                        db
+                                        nil
+                                        (fn [^Connection conn]
+                                          ;; sleep long enough to make sure the PUT request below finishes processing,
+                                          ;; including any async operations that it might trigger
+                                          (Thread/sleep 1000)
+                                          ;; test the connection is open by executing a query
+                                          (try
+                                            (let [stmt      (.createStatement conn)
+                                                  resultset (.executeQuery stmt "SELECT 1")]
+                                              (.next resultset))
+                                            (catch Exception _e
+                                              false)))))]
+          ;; 3. update the database's `database-enable-actions` setting
+          (mt/user-http-request :crowberto :put 200 (format "database/%d" (u/the-id db))
+                                {:settings {:database-enable-actions true}})
+          ;; 4. test the connection was still open at the end of it of the long running process
+          (is (true? @connections-stay-open?))
+          (tx/destroy-db! driver/*driver* empty-dbdef))))))
 
 (deftest fetch-database-metadata-test
   (testing "GET /api/database/:id/metadata"

--- a/test/metabase/api/database_test.clj
+++ b/test/metabase/api/database_test.clj
@@ -523,7 +523,7 @@
       (let [;; 1. create a database and sync
             database-name      (name (gensym))
             empty-dbdef        {:database-name database-name}
-            _                  (tx/create-db! :postgres empty-dbdef)
+            _                  (tx/create-db! driver/*driver* empty-dbdef)
             connection-details (tx/dbdef->connection-details driver/*driver* :db empty-dbdef)
             db                 (first (t2/insert-returning-instances! :model/Database {:name    database-name
                                                                                        :engine  (u/qualified-name driver/*driver*)

